### DISCUSSION TITLE
SPARK MINI

### DISCUSSION THREAD

**Discussion by Gwegb** [soundshed/soundshed-app#57] [Q&A] [answered]
First , I don't know anything about coding so I would like to thank all who have contibuted to this amazing app.
My question is , as they are almost the same ( mini and 40) will there be an update to make soundshed work with SPARK MINI?
Mine is recognized , can change the tone ( switch between 1 2 3 4) but can not see the control chain.

Thank you

**Comment by webprofusion-chrisc** [comment]
I'd like to, I just don't currently own a mini. It's great you can connect and change channels, we just need to figure out what's different about the preset info.

**Reply by Gwegb** [reply]
Yes , certainly, I have the same result on both windows app and web browser app.

**Reply by jxw** [reply]
I would like to help, too!  I just got a Mini and I'm also a software engineer, but for some reason it will not connect to my PC over Bluetooth, only my Android devices :(  If I can fix that, then it would be great to help out.

**Reply by jxw** [reply]
Hmm, I cloned the repo, built the electron version, and was able to connect to my amp right away!  I don't know what the difference is; it wasn't working when I tried the online version, but the electron version works great.

**Reply by jxw** [reply]
Ah, the browser I was using didn't support Bluetooth.  Works in Chrome (in both the web and electron versions, you can change channels, but no controls appear on the screen).

**Reply by webprofusion-chrisc** [reply]
Doh! I bought a mini for Christmas and was waiting until then to investigate :) glad/surprised it's working but I'll get to play with it properly soon.

The issue with controls not showing probably means it's failing to read the preset information (which is a special binary format).

**Reply by webprofusion-chrisc** [reply]
On the topic of Bluetooth in the browser, it would be good to add proper UI telling the user that their browser doesn't support stuff, rather than silently failing as it currently does. Firefox recently added Webmidi but Bluetooth may not happen as they have some devs who are very adamant that Bluetooth is a security risk (the classic example is secretly rewriting your device firmware so it becomes a keylogger, which unfortunately is a very compelling made up argument).

**Reply by jxw** [reply]
> The issue with controls not showing probably means it's failing to read the preset information (which is a special binary format).

Yes, looks like that's what is happening.  I can see in the debug console log messages that Soundshed is sending a request to the Mini for preset info, and the Mini sends back several hundred bytes of data in the [RECV]: messages.  How did you find out how to interpret this data for the Spark 40?  It must be a different format for the Mini

**Reply by jxw** [reply]
Did a little sleuthing today.  Looking at the debug output when Soundshed requested info about preset 0, it looked like this:

Completed preset query 
[SparkDeviceManager]: [RECV]: f0013a120301200f001900005924003631463032 
[SparkDeviceManager]: [RECV]: 464500392d444637382d00343142342d4231f7f0 
[SparkDeviceManager]: [RECV]: 013a470301000f011941362d4300453646373345 
[SparkDeviceManager]: [RECV]: 331035313536265248105954484d23302ef7f001
[SparkDeviceManager]: [RECV]: 3a550301300f02193720
[SparkDeviceManager]: [RECV]: 286900636f6e2e706e67654a42700000172e0062 
[SparkDeviceManager]: [RECV]: 6961732e6e6ff7f0013a390301000f0319697365 
[SparkDeviceManager]: [RECV]: 6758617465421300115d4a3d756e4301114d4a3e 
[SparkDeviceManager]: [RECV]: 292f120211f7f0013a340301080f04194a000000 
[SparkDeviceManager]: [RECV]: 02002a436f6d70726065 
[SparkDeviceManager]: [RECV]: 73736f7243127600114a3e7e3b10f7f0013a2003 
[SparkDeviceManager]: [RECV]: 01300f051901114a3f0874000027426f6f307374 
[SparkDeviceManager]: [RECV]: 657242130003114a3f0f1f7801f7f0013a0b0301 
[SparkDeviceManager]: [RECV]: 180f0619114a000018000002114a000004000024 
[SparkDeviceManager]: [RECV]: 5477696e1b431500114a 
[SparkDeviceManager]: [RECV]: 3f34f7f0013a6e0301500f07194f4a0111454a3e 
[SparkDeviceManager]: [RECV]: 3e51660211594a3f0360640311494a3f2b785204
[SparkDeviceManager]: [RECV]: 11f7f0013a5b0301680f08194a3e3138020f2c43 
[SparkDeviceManager]: [RECV]: 686f72750073416e616c6f671b421400114a3f42 
[SparkDeviceManager]: [RECV]: f7f0013a750301480f09 
[SparkDeviceManager]: [RECV]: 1966250111554a3e0f2f3a0211494a3f3e6a4e03 
[SparkDeviceManager]: [RECV]: 11214a3f115b1d2a44f7f0013a350301000f0a19 
[SparkDeviceManager]: [RECV]: 656c617960526532303143152600114a3e636716 
[SparkDeviceManager]: [RECV]: 1601114a3e1b2012f7f0013a4c0301300f0b1902 
[SparkDeviceManager]: [RECV]: 114a3e3107556c03114a 
[SparkDeviceManager]: [RECV]: 3e3139551a04114a3f090000002b626961f7f001
[SparkDeviceManager]: [RECV]: 3a070301000f0c19732e72653076657262431700
[SparkDeviceManager]: [RECV]: 03114a3f010761013b114a3e46745f02f7f0013a 
[SparkDeviceManager]: [RECV]: 0d0301580f0d19114a3e601b173203114a3f315b
[SparkDeviceManager]: [RECV]: 162004114a3e795b797a
[SparkDeviceManager]: [RECV]: 05114a3e6ef7f0013a420301480f0e0a4a380611 
[SparkDeviceManager]: [RECV]: 1d4a3e19191a17f7 
Connected, refreshing preset..

I noticed that it's mostly 7-bit data, but broken up into chunks with a header byte of 0xf0 and a terminator byte of 0xF7.  Rearranging:

f0013a120301200f001900005924003631463032464500392d444637382d00343142342d4231f7
f0013a470301000f011941362d4300453646373345331035313536265248105954484d23302ef7
f0013a550301300f02193720286900636f6e2e706e67654a42700000172e00626961732e6e6ff7
f0013a390301000f03196973656758617465421300115d4a3d756e4301114d4a3e292f120211f7
f0013a340301080f04194a00000002002a436f6d7072606573736f7243127600114a3e7e3b10f7
f0013a200301300f051901114a3f0874000027426f6f307374657242130003114a3f0f1f7801f7
f0013a0b0301180f0619114a000018000002114a0000040000245477696e1b431500114a3f34f7
f0013a6e0301500f07194f4a0111454a3e3e51660211594a3f0360640311494a3f2b78520411f7
f0013a5b0301680f08194a3e3138020f2c43686f72750073416e616c6f671b421400114a3f42f7
f0013a750301480f091966250111554a3e0f2f3a0211494a3f3e6a4e0311214a3f115b1d2a44f7
f0013a350301000f0a19656c617960526532303143152600114a3e6367161601114a3e1b2012f7
f0013a4c0301300f0b1902114a3e3107556c03114a3e3139551a04114a3f090000002b626961f7
f0013a070301000f0c19732e7265307665726243170003114a3f010761013b114a3e46745f02f7
f0013a0d0301580f0d19114a3e601b173203114a3f315b162004114a3e795b797a05114a3e6ef7
f0013a420301480f0e0a4a3806111d4a3e19191a17f7

Each chunk also has some structure to it:
Bytes 0-2 are always F0 01 3A
Byte 3 varies
Bytes 4-5 are always 03 01
Byte 6 varies
Byte 7 is always 0F
Byte 8 counts up from 00 for each chunk
Byte 9 is (maybe # of bytes to follow, but seems to be slightly off)
Bytes 10-?? payload
Last byte F7

Don't know what it means, but maybe this will help you parse it

**Reply by webprofusion-chrisc** [reply]
The format is based on the work that Paul Hamshere and others did to decode the communications: https://github.com/paulhamsh/Spark-Bluetooth-Message-Format and https://github.com/paulhamsh/Spark

Our TypeScript based message parser is a port of https://github.com/paulhamsh/Spark-Parser/tree/main/SparkClassNew

It's very likely there is just a small change, Paul did mention that the structure was dropping something from the message format on the Spark Programming facebook group: https://www.facebook.com/groups/772009393740813

> 
> for the Spark 40 all data is 7 bitted then wrapped in chunks and then wrapped in blocks. Chunks have headers and end in 0xf7. Blocks have headers and a length in the header.
> The Mini doesn’t wrap with a block - so just has a chunk / data structure

**Reply by webprofusion-chrisc** [reply]
So our TypeScript https://github.com/soundshed/soundshed-app/blob/main/src/spork/src/devices/spark/sparkMessageReader.ts#L112

Should be roughly the equivalent to the python version  https://github.com/paulhamsh/Spark/blob/main/SparkClass/SparkReaderClass.py 

However I don't know if that python code has been adapted to the mini

**Reply by webprofusion-chrisc** [reply]
I also see a comment about dropping the block wrapper but I haven't looked into what the meaning of that is:

>  You can drop the first 16 bytes I think.

**Reply by webprofusion-chrisc** [reply]
A normal Spark 40 preset read would look like `01fe000053fead000000000000000000f0013a150....`

Perhaps a Spark Mini preset is just `f0013a150....`, so skipping the attempt to parse that first part will probably help

**Reply by jxw** [reply]
OK, I've figured out quite a bit more about the Mini preset format, and it's similar to what Soundshed is expecting based on a look at read_preset(), but also different in some significant ways.  It's not too complicated to explain, but might be better if I can send you a more elaborate description in a Word document.

In the meantime, here are some observations:
- Data is sent in 39-byte chunks (the last one might be shorter)
- Each chunk begins with 0xf0 and ends with 0xf7, and all bytes in between are < 0x80
- after the 0xf0, there are a few more header bytes, then the payload
- the 7th, 15th, 23rd and 31st bytes seem to be garbage or maybe partial checksum bytes.  I don't know what they are for, but you have to remove them before you can read the payload bytes out of the remaining data
- If you take the remaining payload bytes, the format is pretty close to what read_preset() expects, with a few differences that I can describe in detail in the Word doc

**Reply by jxw** [reply]
ah, I didn't notice your previous comments - I should take a look at Paul's github page.  Maybe I can just comment there about what I'm seeing in terms of differences in the Mini format

**Reply by webprofusion-chrisc** [reply]
It would be good if his SparkReader class could be updated to read spark mini (which I assume it isn't) but we also need to decide how to read presets from either amp (e.g. detect and discard the header or seek the start of the message).

**Reply by jxw** [reply]
Just read Paul Hamshere's Bluetooth message format document.  Now I understand what that "extra" byte that occurs every 8th byte is for - it holds the uppermost bits of the other 7 bytes, since they were set to 0 in order to make all of the bytes 7-bit data.

So I guess the workflow would be:
- Capture stream of bytes from the amp as is currently done
- Identify the first "0xf0-0xf7" chunk and parse it, because every chunk contains a byte that tells you the total number of chunks to expect
- Get the rest of the chunks based on total number of chunks (as opposed to looking for a final 0xf7)
- Strip off the 0xf0 and 0xf7 bytes
- Convert the remaining 7-bit data back to 8-bit
- Extract just the payload bytes and process them with the Mini-compatible version of read_preset()

**Reply by jxw** [reply]
I'm now pretty sure that the Mini's format is exactly as Paul said, in that it's just the chunks with the block header removed.  The one extra detail I would add is that the first 3 bytes of the data section in the chunk contain:
- total number of chunks in this message
- the number of this chunk (0-based)
- the number of bytes in the rest of the data section (normally 25, but the last chunk is less)

Everything else about the format seems to line up with the protocol description

I agree that we should be able to read presets from either amp.  Maybe that could be based on whether the first few bytes of the received message are a block header or a chunk header?  If it starts with 0x01FE0000, treat it like a Spark 40 preset, otherwise try to parse it as a Spark Mini preset

**Reply by jxw** [reply]
lol - I guess you're way ahead of me - I see that sparkMessageReader.structure_data() already does everything I was thinking of, the only difference being that it removes the block header.  I wonder if supporting the Mini would be as simple as skipping that step for preset info?

**Reply by jxw** [reply]
Progress!
![image](https://user-images.githubusercontent.com/6859385/208883131-f89c6352-2b18-4d28-954c-e2198d3acda5.png)

**Reply by webprofusion-chrisc** [reply]
Great! it's possible to step through the reader in the VS Code debugger or in Chrome dev tools if you edit webpacke.pweb.config.js and enable development mode and inline source maps:

```
mode: 'development',
devtool: 'inline-source-map',
```

Yes the solution may be checking the first part of the data to see if it's a block header (spark 40) or just the preset data (spark mini) and skip the extra work accordingly. The code is very much kept so that it can be easily compared to the SparkReader python code and I can't say I fully understand it having ported it fairly verbatim.

**Reply by jxw** [reply]
I've almost got it!  Works for Rhythm, Lead and Custom, but not Solo for some reason.  I think it's really close, though:
![image](https://user-images.githubusercontent.com/6859385/209000078-8d1e0eab-a13f-47d5-88d5-1b8b27a9595a.png)

**Reply by Gwegb** [reply]
Yeah, great job

**Reply by jxw** [reply]
The only significant change I made was to check the first two bytes to see if it's a block header.  If it is, it works as before.  Otherwise, assume the block header isn't there (that's the Mini format):
![image](https://user-images.githubusercontent.com/6859385/209000834-e052d27b-2c30-4b5b-901b-472d3e21bf33.png)

**Reply by jxw** [reply]
lol - I changed the name of the default Solo preset to Solomon in the Spark App, and now Soundshed can open it:
![image](https://user-images.githubusercontent.com/6859385/209002733-7f97a6cc-2604-4985-b31a-5d522151f9e6.png)

**Reply by webprofusion-chrisc** [reply]
Awesome work! Love that it's just a small change, if you can submit a Pull Request for this that would be great.

Intrigued with the "Solo" issue. I remember there was something about short strings and long strings in the messagepack format or the parsing in general, could be something there needs looked at.

**Reply by jxw** [reply]
OK, pull request submitted.  I apologize if I didn't do it correctly - the interface is a little different than what I'm used to (I normally use Visual Studio 2022).  Hopefully won't be too much of a hassle if I messed up!  In any case, the changes are pretty localized.

**Reply by webprofusion-chrisc** [reply]
Cool that's merged now and will be published soon.

**Comment by Grokka** [comment]
Excellent! Thanks for staying at this.
________________________________

**Reply by Gwegb** [reply]
Yeah guys, you rock

**Comment by webprofusion-chrisc** [comment, marked as answer]
The app (web and desktop versions) now has this update available.

**Reply by Gwegb** [reply]
Well done Guys, great job, thank you , I’m gonna try it and let a feedback

**Reply by webprofusion-chrisc** [reply]
Special thanks to @jxw who did the real work, I just did the corresponding Spark 40 testing and release builds. Ironically (?) the app is not working on my main development Windows machine due to some bluetooth thing, but it worked when I restarted in macOS instead.

**Reply by jxw** [reply]
just tried the online version - works great!  :)

**Comment by Gwegb** [comment]
Really good job Gentlemen , many thanks, it works perfectly with my Spark mini

**Reply by webprofusion-chrisc** [reply]
Great! I've been playing with it as well but I'm not too happy with the reliability, I think we're dropping messages due to the current way messages are handled , so I'm currently rewriting all of how that part is handled in the hope we'll get better reliability. 

I also found that occasionally communication breaks completely and I have to unpair the Spark from windows (forget the device) in order to get communication to work again. That's obviously a windows/chrome things we can't influence.

**Reply by jxw** [reply]
That's great - I hope you figure it out!  Is it specific to the Spark Mini or does the Spark 40 also have issues?  Maybe it's just my luck, but I have also had issues connecting to the Spark app as well (not just Soundshed).

**Reply by webprofusion-chrisc** [reply]
It's kind of specific to the Mini, it's messages seems to be in much smaller chunks and the message terminator 0xf7 is appearing mid-stream, when I try to splice it all back together it's still not coming out as a valid message, so there's more to do there. I think the fact it kind of works for the mini is just luck as to whether the message stream you get ends in 0xf7 or not.

**Reply by jxw** [reply]
Let me know if I can help out!  From looking at the response to the get preset command, it looks like the Mini sends 39-byte chunks that start with 0xf0 and end with 0xf7, except for the last one, which can be anywhere from 12 to 39 bytes.  When Soundshed had trouble reading the Solo preset, it looked like it missed two entire chunks - I couldn't tell if it was due to a flaky connection or a bug in the code that parses the bytes.  Maybe you're having similar issues?

**Reply by jxw** [reply]
One thing that may help with reassembling the payload bytes - the 8th byte of each chunk is the number of chunks in the message, and the 9th byte is the index of the current chunk (0-based).

**Reply by jxw** [reply]
Looking at the debug log for receiving the Solo preset, looks like SparkDeviceManager receives 20, 20, 20, 20 and 10 bytes at a time.  Not sure why.  Looks like it doesn't receive the last bit of the preset because it's less than 20 bytes - changing the preset name to "Solomon" made the last bit long enough that it was received.  So there's something funky with the code that reads in bytes where if the last batch of bytes (NOT the same as chunks) is not long enough, then the software waits forever for the extra bytes that never come.

[SparkDeviceManager]: [RECV]: f0013a6f0301200f001900025924004332304438
[SparkDeviceManager]: [RECV]: 454600452d464638322d00343936412d3939f7f0
[SparkDeviceManager]: [RECV]: 013a720301000f011942312d4300394335463132
[SparkDeviceManager]: [RECV]: 42103042424524534f444c4f23302e3720f7f001
[SparkDeviceManager]: [RECV]: 3a3c0301080f02192869
[SparkDeviceManager]: [RECV]: 636f206e2e706e674a4219700000172e62690061
[SparkDeviceManager]: [RECV]: 732e6e6f6973f7f0013a590301000f0319656761
[SparkDeviceManager]: [RECV]: 743665431300114a3e36422c6901114a00300000
[SparkDeviceManager]: [RECV]: 0002114a3ff7f0013a490301480f04190000002a
[SparkDeviceManager]: [RECV]: 00436f6d707265735873
[SparkDeviceManager]: [RECV]: 6f7242120011554a3e5c10660111f7f0013a0403
[SparkDeviceManager]: [RECV]: 01080f05194a3f36710375294f76657264307269
[SparkDeviceManager]: [RECV]: 76654313002b114a3e483e6901f7f0013a0a0301
[SparkDeviceManager]: [RECV]: 580f0619114a3e4a186d2e02114a3f6604164d25
[SparkDeviceManager]: [RECV]: 506c6578366943150011
[SparkDeviceManager]: [RECV]: 4a3ff7f0013a700301000f07192a50660123114a
[SparkDeviceManager]: [RECV]: 3f262b76023b114a3e681e690313114a3f224312
[SparkDeviceManager]: [RECV]: 04f7f0013a160301180f0819114a3f0004000127
[SparkDeviceManager]: [RECV]: 556e69565869626542130011414a3f0000000111
[SparkDeviceManager]: [RECV]: f7f0013a200301280f09
[SparkDeviceManager]: [RECV]: 194a3f00000c0002114a3f333302332944656c61
[SparkDeviceManager]: [RECV]: 79304d6f6e6f431500f7f0013a540301580f0a19
[SparkDeviceManager]: [RECV]: 114a3e2118562c01114a3e755a011202114a3e5f
[SparkDeviceManager]: [RECV]: 1a043603114a3f19f7f0013a220301580f0b1919
[SparkDeviceManager]: [RECV]: 1a0411214a000000002b
[SparkDeviceManager]: [RECV]: 62006961732e7265765865726243180011f7f001
[SparkDeviceManager]: [RECV]: 3a2b0301280f0c194a3e7f5d4c2e01114a3f2451
[SparkDeviceManager]: [RECV]: 0d4202114a3f3a1d4c2c03114a3e3306f7f0013a
[SparkDeviceManager]: [RECV]: 050301600f0d195904114a623e70674305114a6e
[SparkDeviceManager]: [RECV]: 3e5c2c4606114a6c3f19

Here's what happens when I made the preset name longer:

[SparkDeviceManager]: [RECV]: f0013a6f0301200f001900025924004332304438
[SparkDeviceManager]: [RECV]: 454600452d464638322d00343936412d3939f7f0
[SparkDeviceManager]: [RECV]: 013a720301000f011942312d4300394335463132
[SparkDeviceManager]: [RECV]: 42103042424524534f444c4f23302e3720f7f001
[SparkDeviceManager]: [RECV]: 3a3c0301080f02192869
[SparkDeviceManager]: [RECV]: 636f206e2e706e674a4219700000172e62690061
[SparkDeviceManager]: [RECV]: 732e6e6f6973f7f0013a590301000f0319656761
[SparkDeviceManager]: [RECV]: 743665431300114a3e36422c6901114a00300000
[SparkDeviceManager]: [RECV]: 0002114a3ff7f0013a490301480f04190000002a
[SparkDeviceManager]: [RECV]: 00436f6d707265735873
[SparkDeviceManager]: [RECV]: 6f7242120011554a3e5c10660111f7f0013a0403
[SparkDeviceManager]: [RECV]: 01080f05194a3f36710375294f76657264307269
[SparkDeviceManager]: [RECV]: 76654313002b114a3e483e6901f7f0013a0a0301
[SparkDeviceManager]: [RECV]: 580f0619114a3e4a186d2e02114a3f6604164d25
[SparkDeviceManager]: [RECV]: 506c6578366943150011
[SparkDeviceManager]: [RECV]: 4a3ff7f0013a700301000f07192a50660123114a
[SparkDeviceManager]: [RECV]: 3f262b76023b114a3e681e690313114a3f224312
[SparkDeviceManager]: [RECV]: 04f7f0013a160301180f0819114a3f0004000127
[SparkDeviceManager]: [RECV]: 556e69565869626542130011414a3f0000000111
[SparkDeviceManager]: [RECV]: f7f0013a200301280f09
[SparkDeviceManager]: [RECV]: 194a3f00000c0002114a3f333302332944656c61
[SparkDeviceManager]: [RECV]: 79304d6f6e6f431500f7f0013a540301580f0a19
[SparkDeviceManager]: [RECV]: 114a3e2118562c01114a3e755a011202114a3e5f
[SparkDeviceManager]: [RECV]: 1a043603114a3f19f7f0013a220301580f0b1919
[SparkDeviceManager]: [RECV]: 1a0411214a000000002b
[SparkDeviceManager]: [RECV]: 62006961732e7265765865726243180011f7f001
[SparkDeviceManager]: [RECV]: 3a2b0301280f0c194a3e7f5d4c2e01114a3f2451
[SparkDeviceManager]: [RECV]: 0d4202114a3f3a1d4c2c03114a3e3306f7f0013a
[SparkDeviceManager]: [RECV]: 050301600f0d195904114a623e70674305114a6e
[SparkDeviceManager]: [RECV]: 3e5c2c4606114a6c3f19
[SparkDeviceManager]: [RECV]: f0014d3b0371381103014c7f4d0e030c4d0d3817
[SparkDeviceManager]: [RECV]: f7

That works

**Reply by webprofusion-chrisc** [reply]
Thanks! If you look at the comms-wip branch I've update the ble message receiver to push data received into a queue rather than trying to handle it. My theory was we were losing something due to taking too long to handle the messages. I've experimented with splitting the messages by the f7 terminator as they come in so the message queue only ever contains full messages.

This seems to work theoretically (there is a test in the settings page) but in reality something is still not quite right. Feel free to have a look, I'll work on it sometime over the next couple of weeks.

**Reply by webprofusion-chrisc** [reply]
It's probably going to be useful to reduce the code down to a python BLE example and a javascript bluetooth example, just to confirm that both are reading the same raw data, then (for javascript) feed the data to our reader and make sure it parses into the data into the expected messages. It would be interesting to see if the simplest raw data received by python is different to javascript as this would imply a difference in the BLE implementation

Our original code was built to use bluetooth serial via electron (raw read from a serial socket) then modified to use either electron or web (ble, receiving change data from a GATT characteristic), then finally changed to only use BLE.

**Reply by jxw** [reply]
Happy New Year!

I worked on this for the last couple of days and I've gotten to a point where the behavior is much improved.  Seems to work 100% for any commands that expect a single chunk reply, and most of the time (95+%?) for multi-part messages like presets.  On some occasions, it doesn't seem to get all of the chunks.  Retrying seems to work.  Anyway, I can send you a pull request to take a look at the changes

**Reply by jxw** [reply]
Hi, sorry, I tried to create a pull request but had issues.  :(

Here are the files I touched so you can take a look.  Basically, my intent was to assemble the received bytes into complete chunks _before_ pushing them into the receive queue, rather than after.
[spork_src_devices_spark.zip](https://github.com/soundshed/soundshed-app/files/10333302/spork_src_devices_spark.zip)

**Reply by webprofusion-chrisc** [reply]
Thanks! I'll give this a try later on, very much appreciated. 

I think we're both trying something similar in different ways, your trick of skipping the message if it doesn't have the starting marker is interesting. I notice you mentioned handling chunks that are out of order - I spent a little time looking at whether we may sometimes be receiving data out of order but I haven't really figured out if that's true or not. It does appear that sometimes we just don't get a full response from the amp (which could either be a bug in Chromium BLE or the amp, either way there's nothing we can do with those).

**Reply by webprofusion-chrisc** [reply]
Gave it a try, it definitely works better for querying the preset! 

Overall though I'm not confident that we currently have a well understood model for how to work with both the spark 40 and the mini and I'm starting to feel like we're writing increasingly clever code to work around a fundamental communications issue. 

I'm going to take a break from it then try to take this back to basics by capturing a larger set of send/receive messages then build tests to run through those. Ideally I'd like to end up with the simplest possible code for reliably handling these messages. which I think is a mixture of the current approaches.

**Reply by webprofusion-chrisc** [reply]
Further to this I've implemented a basic python and js test (e.g. https://github.com/soundshed/spork/tree/master/examples/js) for raw communication where it just loops through the presets and queries them. The results are byte-for-byte identical between JS and Python so there is no ordering issue at the browser level. 

I _believe_ the root of the reliability problem is not allowing enough time for acknowledgements/replies before executing further commands, so I'm going to implement something to ensure the previous operation is fully completed before moving to the next.

**Reply by jxw** [reply]
Awesome - thanks for really digging into this!  I think comms-wip head is not that far away from still supporting both the 40 and the Mini - the main thing is to make accumulateChunk() in bleProvider.ts smart enough to know what to do with block headers.

I agree with you that the reliability problem feels like a timing issue.  Also, for reasons I don't understand, it looks like we're occasionally receiving the preset data more than once per read preset command - we get a complete message plus the same preset repeated, but only the first few chunks.  Do we send the read preset command more than once? I have to go back and take a closer look at the logs, but I don't think so; maybe the response is just the weird way the Mini works (does it just keep sending the preset data until the app sends another command?  No idea).

**Reply by jxw** [reply]
By the way, the reason I implemented the code to handle out of order chunks wasn't because I suspect they could be coming in a random order, but to handle cases where we might get something like chunk 3, 4, 5,...15, 1, 2,... which I did see happening.  If you can figure out the timing issue so we never miss the first chunk, then we could simplify the implementation accordingly, which would be way better.

**Reply by webprofusion-chrisc** [reply]
Thanks, yep have done further experiments and the root cause of weirdness was indeed clunky timing. Tracking when the last msg was received (or last CMD issued) and simply preventing the next CMD until some time has elapsed fixes most of it. We both built something to take the byte stream and terminate it/queue it as it comes in so I'm working through what can be simplified or removed etc. Just working on getting the spark 40 bit to behave.

**Reply by jxw** [reply]
Latest comms-wip seems pretty solid when it comes to any single-chunk messages, like channel changes, effect on/off, etc.  Reading presets is still a bit iffy - not sure why.  Progress, though!  (speaking for Mini only - don't have a 40)

**Reply by webprofusion-chrisc** [reply]
Agreed, we still seem to get the occasional incomplete preset back in the raw ble.

I think perhaps the last piece of the puzzle would be to match CMDs with ACKs, and if we don't fully resolve the reply to a command after enough time then we repeat the command.

Spark 40 comms definitely seems more reliable than the Spark Mini, not sure why. The Spark Mini also switches channel much slower than the 40 (even just using the amp control), so could perhaps be the CPU maxing out.

**Reply by Gwegb** [reply]
Have an other question, in the setting tab of the app, we can chose a MIDI input to control Spark, but none of my controller appear when I click ( only a small white windows), so my question is, do you know how the spark control ( or the Xsonic airstep) communicate with the Spark? Is it MIDI or an other Protocol?

**Reply by webprofusion-chrisc** [reply]
I think some of the footpedals actually send bluetooth commands to the spark app (on android or ios) rather than directly to the amp, possibly some of them talk directly to the amp though (same way we do). I don't think it's midi to the amp as that would have been a lot easier than the communication we do!

Regarding MIDI in the soundshed app, we use the Web Midi feature, which for me just lists connected controllers (via USB). Which type of controller do you have? I've tried soundshed successfully with a Korg Microkey keyboard and a Line 6 HD 400 controller. (via Chrome, on Windows).

**Reply by webprofusion-chrisc** [reply]
If you got to this app: https://midisandbox.com/ and click Start Playing, then choose your midi input, does it show up? It will only work in Chrome or Edge.

**Reply by Gwegb** [reply]
Thank you Christopher,

I tried midisandbox and my Ampero control does appear there ( I use Firefox ) but does not in Soundshed ( web or app ).
Anyway your probably right , I connected Spark Mini with the iOS app and I used MIDI FLOW to see if there was a midi communication via bluetooth with the IPAd but nothing happened when BIAS FX clearly has MIDI control.

**Reply by webprofusion-chrisc** [reply]
Thanks, Firefox isn't supported by the soundshed app as it doesn't have Bluetooth support at all (for speaking to the spark amp) and Firefox only recently enabled Web Midi via a permission add-on that's very random/unstable (I can get it to work with some sites, but it doesn't offer to work with soundshed). I've had fairly heated arguments with some of the security gatekeepers affiliated with Mozilla regarding bluetooth, usb etc - they are extermely hostile to such things. Their basic premise is that users are too stupid to be allowed to have web usb/bluetooth because they'll visit random sites that re-write their device firmware and chaos will ensue.  You can basically only use Google Chrome or Microsoft Edge.

**Reply by webprofusion-chrisc** [reply]
Just as a follow up I just bought a new midi controller and can confirm the midi mapping is not working properly currently. If you have a pre-existing saved midi controller it works but not if you haven't set it yet. Working on it.

**Comment by webprofusion-chrisc** [comment]
The new Store Settings option has now been added to the web version to send a preset to the current channel.
